[bugfix] missing supported channels

itayx opened the pull request:
When generating the Basics Radio Capabilities, some of the channels that should be supported are not preset. The channels are retrieved from the NL correctly as they appear in the log print of the ap_manager_thread.

But before they are processed by the `add_ap_radio_basic_capabilities` into the WSC autoconfig radio basic capabilities TLV, they are cut short due to the size limiter `SUPPORTED_CHANNELS_LENGTH` which is set to 64.

To fix this issue, the value of `SUPPORTED_CHANNELS_LENGTH` need to be increased

itayx committed:
son_slave: fix initialization

This is a preparative commit.

The supported channels vector in the son_slave need to be cleared out
when receiving the AP_JOINED message in case the notification message
will be sent more then once.
The supported channels array need to be filled with blank instances of
the sWifiChannel struct so during the `add_ap_radio_basic_capabilities`
no "junk" structures will be mistakenly used for building the basic
radio capabilities TLV.

Clear the vector before inserting the new supported channels.
Initialize the array before copying into it the stored channels.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
beerocks_defines: update SUPPORTED_CHANNELS_LENGTH

The SUPPORTED_CHANNELS_LENGTH is used throughout the code as the limiter
for the supported channels containers. but for some APs the supported
channels list can exceed the 64 channel limit,therefore the limit needs
to be raised.

Increase the value of SUPPORTED_CHANNELS_LENGTH.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

kantera800 commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -43,7 +43,7 @@ enum eStructsConsts {
     NODE_NAME_LENGTH          = 32,
     IFACE_NAME_LENGTH         = 32 + 4, //need extra 1 byte for null termination + alignment
-    SUPPORTED_CHANNELS_LENGTH = 64,     //support upto # channels, every channel item is 32-bit
+    SUPPORTED_CHANNELS_LENGTH = 80,     //support upto # channels, every channel item is 32-bit
```
didn't we agree on 128 ?

itayx replied:
When I expanded the size to 128, it failed the `test_flows` the same way currently the 80 fails with the `test-on-netgear-rax40`

I believe the tests fail due to an issue with the AP_JOINED and M1 messages, as I see (in both cases) that the agent is stuck in a never-ending loop of `STATE_WAIT_FOR_JOINED_RESPONSE`

arnout replied:
That's probably because sNodeHostap becomes too large. With 128 channels, the preferred channel list in it becomes 1K so it's very easy to reach the maximum TLV size.

The best solution would be to remove preferred channels from sNodeHostap completely and instead use the channel preference flow. But that's more work, I guess.

kantera800 replied:
yes we should find a w/a for now.
@itayx  please  check what can we remove from supported_channels for now that is not used.
also go over sNodeHostap parameters and see if there is a something we can remove.
If nothing can be removed i propose removing them from sNodeHostap and create new VS CMDU only from supported channels.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3507,8 +3508,8 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
-            supported_channels_arr;
-        std::copy_n(supported_channels.begin(), supported_channels_arr.size(),
+            supported_channels_arr{{}};
+        std::copy_n(supported_channels.begin(), supported_channels.size(),
```
This is an actual fix, so the "preparative commit" part of the commit message is not really needed. (nitpick, obviously)

arnout reviewed (approved):
Good enough for me. However, does it really solve the issue? IIRC there are actually 100 supported channels...

itayx committed:
tree-wide: move preferred_channels outside the sNodeHostap

The preferred_channels is stored within the sNodeHostap, but since the
increased value of SUPPORTED_CHANNELS_LENGTH, the SLAVE_JOINED_RESPONSE
is never sent from the son_master due to the size of the message.

Move the preferred_channels to outside the sNodeHostap.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
bwl: fix missing preferred channels in nl80211

In the ap_wlan_hal_nl80211::read_preferred_channels the preferred
channels are read from the nl80211_client::get_radio_info, but the
channels are never added to the bwl's radio_info preferred_channels.

This error has gone unnoticed due to the ap manager using a fixed size
list instead of a dynamic one.

Store read channels in the radio_info's preferred_channels.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx commented:
@arnout @kantera800
I've fixed the issue with the message size and fixed an unnoticed bug with `ap_wlan_hal_nl80211::read_preferred_channels`.
I moved the preferred_channels from the `sNodeHostap` to the body of `ACTION_APMANAGER_JOINED_NOTIFICATION` in a dynamically allocated list, then stored the preferred_channels in the son_slave directly (Before it was stored as part of the `sNodeHostap` struct in son_slave).
I've also increased the SUPPORTED_CHANNEL_LENGTH to 128. to it could support twice as many channels as before **(If there is a maximum number of channels that we should use, we should set it to that value)**.
As for the unnoticed bug i mentioned above, the read_preferred_channels never really stored the preferred channels, but this has gone unnoticed because the old fixed-size list would simply use blank sWifiChannel structs.

kantera800 reviewed (approved):
Looks good , not sure why we even sent preffered_channels in the first place to the controller.
